apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs

```diff
@@ -1111,5 +1119,103 @@ public void AckModifyAckDeadlineFault_Thrown([CombinatorialValues(true, false)]
             });
         }
+
+        [Theory, CombinatorialData]
+        public void ExactlyOnce_TemporaryFault([CombinatorialValues(true, false, null)] bool? ackNackOrExtends)
+        {
+            var msgs = new[]
+            {
+                ServerAction.Data(TimeSpan.Zero, new[] { "1" }),
+                ServerAction.Data(TimeSpan.Zero, new[] { "2" }),
+                ServerAction.Data(TimeSpan.Zero, new[] { "3" }),
+                ServerAction.Data(TimeSpan.Zero, new[] { "4" }),
+                ServerAction.Inf()
+            };
+
+            // Any temporary failure in ack/nack/extend request should not be thrown to the client. It should be retried.
+            var exception = new RpcException(new Status(StatusCode.DeadlineExceeded, ""), "");
+            // If ackNackOrExtends is null, then it is extend request. ModifyAcknowledgeDeadline RPC will throw the supplied exception.
+            // If ackNackOrExtends is true, then it is acknowledge request. Acknowledge RPC will throw the supplied exception.
+            // If ackNackOrExtends is false, then it is nack request. ModifyAcknowledgeDeadline RPC will throw the supplied exception.
+            var ackModifyAckDeadlineAction =
+                ackNackOrExtends == null ? AckModifyAckDeadlineAction.Data(exception, 10, false, true)
+                : ackNackOrExtends.Value ? AckModifyAckDeadlineAction.BadAck(exception, 10)
+                : AckModifyAckDeadlineAction.BadModifyAckDeadline(exception, 10);
+
+            using var fake = Fake.Create(new[] { msgs }, ackModifyAckDeadlineAction: ackModifyAckDeadlineAction, isExactlyOnce: true);
+            fake.Scheduler.Run(async () =>
+            {
+                var handledMsgs = new List<string>();
+                var doneTask = fake.Subscriber.StartAsync(async (msg, ct) =>
+                {
+                    handledMsgs.Locked(() => handledMsgs.Add(msg.Data.ToStringUtf8()));
+                    if (ackNackOrExtends == null)
+                    {
+                        // Add delay greater than ackDeadline to simulate extends.
+                        await fake.TaskHelper.ConfigureAwait(fake.Scheduler.Delay(TimeSpan.FromSeconds(70), ct));
+                    }
+
+                    return await Task.FromResult(ackNackOrExtends == false ? SubscriberClient.Reply.Nack : SubscriberClient.Reply.Ack);
+                });
+                await fake.TaskHelper.ConfigureAwait(fake.Scheduler.Delay(TimeSpan.FromSeconds(100), CancellationToken.None));
+                await fake.TaskHelper.ConfigureAwait(fake.Subscriber.StopAsync(CancellationToken.None));
+                // Despite temporary failures, all 4 messages should be handled.
+                Assert.Equal(new[] { "1", "2", "3", "4" }, handledMsgs);
+            });
+        }
+
+        [Theory, CombinatorialData]
+        public void ExactlyOnce_PermanentFault([CombinatorialValues(true, false, null)] bool? ackNackOrExtends)
```

Comment:
Given how much conditional code there is, possibly separate out extend from ack/nack?

Reply:
Done.